rcl/src/rcl/security.c

```diff
@@ -0,0 +1,356 @@
+// Copyright 2018 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "rcl/security.h"
+
+#include <stdbool.h>
+
+#include "rcl/error_handling.h"
+
+#include "rcutils/logging_macros.h"
+#include "rcutils/filesystem.h"
+#include "rcutils/format_string.h"
+#include "rcutils/get_env.h"
+#include "rcutils/strdup.h"
+
+#include "rmw/security.h"
+#include "rmw/security_options.h"
+
+#ifdef __clang__
+# pragma clang diagnostic push
+# pragma clang diagnostic ignored "-Wembedded-directive"
+#endif
+#include "tinydir/tinydir.h"
+#ifdef __clang__
+# pragma clang diagnostic pop
+#endif
+
+rcl_ret_t
+rcl_get_security_options_from_environment(
+  const char * name,
+  const char * namespace_,
+  const rcutils_allocator_t * allocator,
+  rmw_security_options_t * security_options)
+{
+  bool use_security = false;
+  rcl_ret_t ret = rcl_security_enabled(&use_security);
+  if (RCL_RET_OK != ret) {
+    return ret;
+  }
+  RCUTILS_LOG_DEBUG_NAMED(
+    ROS_PACKAGE_NAME, "Using security: %s", use_security ? "true" : "false");
+
+  if (!rmw_use_node_name_in_security_directory_lookup()) {
+    ret = rcl_get_enforcement_policy(&security_options->enforce_security);
+    if (RCL_RET_OK != ret) {
+      return ret;
+    }
+
+    if (!use_security) {
+      security_options->enforce_security = RMW_SECURITY_ENFORCEMENT_PERMISSIVE;
+    } else {  // if use_security
+      // File discovery magic here
+      char * secure_root = rcl_get_secure_root(
+        name,
+        namespace_,
+        allocator);
+      if (secure_root) {
+        RCUTILS_LOG_INFO_NAMED(ROS_PACKAGE_NAME, "Found security directory: %s", secure_root);
+        security_options->security_root_path = secure_root;
+      } else {
+        if (RMW_SECURITY_ENFORCEMENT_ENFORCE == security_options->enforce_security) {
+          return RCL_RET_ERROR;
+        }
+      }
+    }
+  }
+  return RCL_RET_OK;
+}
+
+rcl_ret_t
+rcl_security_enabled(bool * use_security)
+{
+  const char * ros_security_enable = NULL;
+  const char * get_env_error_str = NULL;
+
+  if (!use_security) {
```

Comment:
We have a macro for this: 

`RCL_CHECK_ARGUMENT_FOR_NULL(use_security, RCL_RET_INVALID_ARGUMENT)`